src/allocator.rs

```diff
@@ -4,8 +4,52 @@ use crate::reduction::EvalErr;
 use chia_bls::{G1Element, G2Element};
 use clvm_traits::{ClvmDecoder, ClvmEncoder, FromClvmError, ToClvmError};
 
+const MAX_NUM_ATOMS: usize = 62500000;
+const MAX_NUM_PAIRS: usize = 62500000;
+const NODE_PTR_IDX_BITS: u32 = 26;
+const NODE_PTR_IDX_MASK: u32 = (1 << NODE_PTR_IDX_BITS) - 1;
+
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct NodePtr(pub i32);
+pub struct NodePtr(pub u32);
+
+enum ObjectType {
+    Pair,
+    Bytes,
+}
+
+// The top 6 bits of the NodePtr indicate what type of object it is
+impl NodePtr {
+    fn new(t: ObjectType, idx: usize) -> Self {
+        assert!(idx <= NODE_PTR_IDX_MASK as usize);
+        NodePtr(((t as u32) << NODE_PTR_IDX_BITS) | (idx as u32))
+    }
+
+    fn node_type(&self) -> (ObjectType, usize) {
+        (
+            match self.0 >> NODE_PTR_IDX_BITS {
+                0 => ObjectType::Pair,
+                1 => ObjectType::Bytes,
+                _ => {
+                    panic!("unknown NodePtr type");
+                }
+            },
+            (self.0 & NODE_PTR_IDX_MASK) as usize,
+        )
+    }
+
+    pub fn to_index(&self) -> usize {
```

Comment:
Does this need to be fully public? Perhaps `pub(crate)` would suffice here?

Reply:
I also renamed this to `as_index()` to please clippy. However, I believe this function won't live for very long